syz-ci/jobs.go

```diff
@@ -26,72 +27,146 @@ import (
 	"github.com/google/syzkaller/vm"
 )
 
+type JobManager struct {
+	cfg             *Config
+	dash            *dashapi.Dashboard
+	managers        []*Manager
+	shutdownPending <-chan struct{}
+}
+
 type JobProcessor struct {
 	cfg             *Config
 	name            string
+	instanceSuffix  string
 	managers        []*Manager
 	knownCommits    map[string]bool
-	stop            chan struct{}
-	shutdownPending chan struct{}
 	dash            *dashapi.Dashboard
-	syzkallerRepo   string
-	syzkallerBranch string
+	baseDir         string
+	jobFilter       *ManagerJobs
+	jobTicker       <-chan time.Time
+	commitTicker    <-chan time.Time
+	shutdownPending <-chan struct{}
 }
 
-func newJobProcessor(cfg *Config, managers []*Manager, stop, shutdownPending chan struct{}) (*JobProcessor, error) {
+func newJobManager(cfg *Config, managers []*Manager, shutdownPending chan struct{}) (*JobManager, error) {
 	dash, err := dashapi.New(cfg.DashboardClient, cfg.DashboardAddr, cfg.DashboardKey)
 	if err != nil {
 		return nil, err
 	}
-	return &JobProcessor{
+	return &JobManager{
 		cfg:             cfg,
-		name:            fmt.Sprintf("%v-job", cfg.Name),
+		dash:            dash,
 		managers:        managers,
-		knownCommits:    make(map[string]bool),
-		stop:            stop,
 		shutdownPending: shutdownPending,
-		dash:            dash,
-		syzkallerRepo:   cfg.SyzkallerRepo,
-		syzkallerBranch: cfg.SyzkallerBranch,
 	}, nil
 }
 
-func (jp *JobProcessor) loop() {
-	jobTicker := time.NewTicker(time.Duration(jp.cfg.JobPollPeriod) * time.Second)
-	commitTicker := time.NewTicker(time.Duration(jp.cfg.CommitPollPeriod) * time.Second)
-	defer jobTicker.Stop()
+// backgroundLoop starts a job loop in parallel and returns a blocking function
+// to gracefully stop job processing.
+func (jm *JobManager) backgroundLoop(wg *sync.WaitGroup) func() {
+	stop := make(chan struct{})
+	done := make(chan struct{}, 1)
+	go func() {
+		jm.loop(stop)
+		done <- struct{}{}
+		wg.Done()
+	}()
+	return func() {
+		stop <- struct{}{}
+		<-done
+	}
+}
+
+func (jm *JobManager) loop(stop chan struct{}) {
+	if err := jm.resetJobs(); err != nil {
+		if jm.dash != nil {
+			jm.dash.LogError("syz-ci", "reset jobs failed: %v", err)
+		}
+		return
+	}
+	commitTicker := time.NewTicker(time.Duration(jm.cfg.CommitPollPeriod) * time.Second)
 	defer commitTicker.Stop()
+	jobTicker := time.NewTicker(time.Duration(jm.cfg.JobPollPeriod) * time.Second)
```

Comment:
This should be per-JobProcessor. Otherwise only 1 of them receives each timer signal.
Or is it intentional?

Reply:
It was intentional - I did not want to further increase the load on the dashboard.